### ISSUE TITLE
Patching Berkeley Mono v2 leads to incorrectly spaced chars

### ISSUE THREAD

**Issue by clementpoiret** [ryanoasis/nerd-fonts#1772]
### Requirements

- [X] I have searched the [issues](https://github.com/ryanoasis/nerd-fonts/issues) for my issue and found nothing related or helpful
- [X] I have searched the [FAQ](https://github.com/ryanoasis/nerd-fonts/wiki/FAQ-and-Troubleshooting) for help
- [X] I have checked the [Wiki](https://github.com/ryanoasis/nerd-fonts/wiki) for help

### Experienced Behavior

The arrow ligature is offcentered to the left, and the `r` in the `import` word is to far to the right. It happens for many characters.
Even if the font is behind a paywall, the issue should also happen on the free trial version of Berkeley Mono v2. Command used to patch fonts:

```
sudo docker run --rm \
  -v /home/clementpoiret/.local/share/fonts/TX-02:/in \
  -v /home/clementpoiret/.local/share/fonts/TX02-NerdFont:/out \
  nerdfonts/patcher \
  --progressbars \
  --mono \
  --adjust-line-height \
  -c
```

### Expected Behavior

Same behavior as for other fonts, centered chars and glyphs.

### Example Symbols or Text

`-> import` but it happens for a lot of characters and nearly all ligatures.

### Font Used

Berkeley Mono v2

### Source of Font File

https://usgraphics.com/products/berkeley-mono

### Terminal Emulator (and the title of the terminal window)

ghostty, kitty

### Operating System and Version

NixOS Unstable

### Screenshots

Unpatched font:
![2024-12-30-At-10h39m30s](https://github.com/user-attachments/assets/7e96bce6-f8f5-4d62-8e70-3136566bf4bd)

Patched font:
![2024-12-30-At-10h37m13s](https://github.com/user-attachments/assets/dc94962f-4751-4039-9c56-445c81d539eb)

**Comment by Finii** [comment]
Sorry to hear of your problems.

Just from your discription I'd say the problem is how the font constructs the ligatures. Probably not as CALT what is customary.
If that is the case there is nothing we can do for the parameters yoy give.

I guess the patched font is ok if you omit `--mono`? What is the reason for that flag, maybe we can solve that differently.

Also `--adjust-line-height`, do you have a specific reason to give that? I in fact wanted to remove that option several times already and I search for ppl to give me reason why to keep it.

Sent from Nine
________________________________
From: Clément POIRET ***@***.***>
Sent: Monday, 30 December 2024 11:05
To: ryanoasis/nerd-fonts
Cc: Subscribed
Subject: [ryanoasis/nerd-fonts] Patching Berkeley Mono v2 leads to incorrectly spaced chars (Issue #1772)

Requirements
I have searched the issues for my issue and found nothing related or helpful
I have searched the FAQ for help
I have checked the Wiki for help
Experienced Behavior
The arrow ligature is offcentered to the left, and the r in the import word is to far to the right. It happens for many characters.
Even if the font is behind a paywall, the issue should also happen on the free trial version of Berkeley Mono v2.
Expected Behavior
Same behavior as for other fonts, centered chars and glyphs.
Example Symbols or Text
-> import but it happens for a lot of characters and nearly all ligatures.
Font Used
Berkeley Mono v2
Source of Font File
https://usgraphics.com/products/berkeley-mono
Terminal Emulator (and the title of the terminal window)
ghostty, kitty
Operating System and Version
NixOS Unstable
Screenshots
Unpatched font:
2024-12-30-At-10h39m30s.png (view on web)
Patched font:
2024-12-30-At-10h37m13s.png (view on web)

**Comment by Finii** [comment]
Ah, I see this person uses the same additional flags, and you just use `-c` (complete) for easiness :+1:

I believe in the demo of TX-02 there are no ligs, as this became an add on buy option? [1]

Anyhow, `--mono` does mess around with the source font in ways that you probably do not need. It forces the whole font to be some specific kind of monosopaced that ALL terminals recognise. When I see the issues that you describe I guess the reason is that the font is forced into being really monospaced.

So you should propably avoid that.

Please try without `--mono`, which will also leave you with bigger icons - which most (?) people prefer anyhow.
If you must have small (within one monospace cell) icons, that would need a script change (a very simple one), maybe we should add an option for that. If we tried and found it is a solution for you. In retrospect `--mono` should have been two options maybe from the start because it does two things. Well, usually you want both. 🤔 

The demo font FX-050 (?) has to be obtained via some fake order 🙄 [2]

Anyhow. It would be good if you could run the patcher with `--debug 2` and post the first lines where the source font analysis is output.
I'm not sure that it works correctly with the docker image, and anyhow it would be good if you either
- enter the docker image into a shell and run there interactively
- or use the FontPatcher.zip without docker on your machine directly with all dependencies installed

In that way I can have a look at the debug output and tweak the font-patcher script to not touch the ligs on patching even with `--mono`.

[1] https://usgraphics.com/catalog/FX-102
[2] https://usgraphics.com/catalog/FX-050

**Comment by Finii** [comment]
Oh, the misaligned letter `r` ... I have no explanation for that. Are more ordinary letters misaligned?
You say you are on NixOS 🤔 There are issues on Windows that could explain it, but none I'm aware of on Nix 😒 
This looks like a Font Family mixup. Maybe try to install only one font file of all the files and try again.
If that solves it we would need to have a look on the naming. You did not specify any `--makegroups`, what you need can be decided on the complete output of `--debug 2` when you patch all fonts in the family that you plan to install in parallel.

**Comment by clementpoiret** [comment]
Thanks for your replies. I'm not in front of my computer right now, but I'll come back with debug logs asap. In the meantime, I have some more details:
- As this is a v2, I simply naively tried commands others used to patch Berkeley Mono v1,
- I tried with and without `--mono` and `--adjust-line-height`, same issue,
- The `r` is not the only char to be moved, I don't know why but it's the most visible one,
- it reminds me an issue I had when trying (an earlier version of) MonaSpice, which had similar ligature issues on my side. I didn't patch this one myself,
- patching the TTF files produces correct fonts, only patching the OTF ones are producing those weird misalignments

**Comment by Finii** [comment]
Hmm, the demo font is kind of useless. It just contains these basic glyphs, and no ligatures or any other stuff.


> patching the TTF files produces correct fonts, only patching the OTF ones are producing those weird misalignments

Hmm, is there a specific reason that you prefer the `otf` variant? The `ttf` is also an OpenType font, nowadays ;) and it has technically more capabilities. If I have a choice here we usually patch (only) the `ttf` variants.

<img width="961" alt="Screenshot 2024-12-31 at 10 12 02" src="https://github.com/user-attachments/assets/cc8e4c59-f551-46be-9129-73c8759d7b82" />

_Just the basic glyphs_

<img width="853" alt="Screenshot 2024-12-31 at 10 22 30" src="https://github.com/user-attachments/assets/46185fde-2827-4243-a6d5-438a701dc9c7" />

_While the `otf` has the blues defined the `ttf` seems to be not even autohinted_

**Comment by mrp52** [comment]
Just ran into this myself. So far I am unable to find a solution playing around with a whole bunch of options to the patcher docker image. You can see the issue very clearly in the two screenshots of fastfetch, things like the "mor" in my name with the o and r being spaced too far apart and r and m in "arm64" being too close together. 

First image is the unpatched (trial) font. Second is patched. 

<img width="1222" alt="trial" src="https://github.com/user-attachments/assets/04c3e0a3-9513-4bf2-bbbd-c9af0ec51efd" />
<img width="1222" alt="patched" src="https://github.com/user-attachments/assets/eae86fd0-7345-46f6-9d1e-da973b94581f" />

**Comment by Finii** [comment]
I believe, what destroys the ligs is this

```python
        if self.args.single:
            # Force width to be equal on all glyphs to ensure the font is considered monospaced on Windows.
            # This needs to be done on all characters, as some information seems to be lost from the original font file.
            self.set_sourcefont_glyph_widths()
```

I will just quickly create a new docker image with an option to suppress that.
But I think that will not solve the 'r' issue. Maybe one step at a time.

Lets see.

Will open a PR now and merge it, maybe we can discuss if this is a viable solution there.

**Comment by Finii** [comment]
Wait, what, it is in the trial font?
Testing myself.

**Comment by mrp52** [comment]
> Wait, what, it is in the trial font? Testing myself.

Yes I am using the trial font (wanted to see if I could patch it before buying it as it isn't cheap!). It is a little annoying you have to "sign up" to get the trial download.

**Comment by Finii** [comment]
Well, see it. Should be easy to find the reason...

<img width="853" alt="Screenshot 2024-12-31 at 14 29 04" src="https://github.com/user-attachments/assets/5d4ba31a-90fd-468b-9abe-7af38feef020" />

**Comment by Finii** [comment]
Hmm, the demo font has this defect on purpose 😬 
This is not added by the patcher script.

<img width="853" alt="Screenshot 2024-12-31 at 14 30 57" src="https://github.com/user-attachments/assets/0c2f300a-16f9-4cc8-b699-a60eb9ed2b1f" />

**Comment by Finii** [comment]
Interestingly can not see that in neither Glyphs3 nor FontGoggles 🤔 

<img width="1470" alt="Screenshot 2024-12-31 at 14 37 36" src="https://github.com/user-attachments/assets/2a505b7b-55c8-446c-ab86-e2cba6341fb2" />

**Comment by Finii** [comment]
The shifted letters like `r` seem to be a fontforge bug.
Built HEAD from git, and it is still there. The glyphs are all somehow 'right aligned'-ish, with some it is most obvious, with some it is not so much. See image below.

This does not happen when the `ttf` is opened, just the `otf` has that.

Also tried to convert the `otf` to `ttx` and back to `otf`, and it is still the same.
While the individual glyphs in ttx look ok, here the vertical bar (where the code is very easy):

```
        <CharString name="bar">
          262 -86 rmoveto
          76 872 -76 hlineto
          endchar
        </CharString>
```

It has 262 left hand side bearing, is 76 wide and with a an advance width of 600 the right hand side bearing calculates also to 262 (600 - 262 - 76).
Strangely this is not how fontforge reads it.

----

![image](https://github.com/user-attachments/assets/c8b7f2f1-8d7f-4d5c-99ad-79971bce6a6a)

_Opened with self-built from HEAD fontforge_

![image](https://github.com/user-attachments/assets/4c8cc138-3d8d-49bc-9141-1c28cda671d9)

_Description of 'bar' as decoded by ttx_


_As opened in Glyphs3_

![image](https://github.com/user-attachments/assets/811f48d0-1b21-45a8-9641-ba71a255f810)


_As opened in fontforge_

**Comment by clementpoiret** [comment]
Thanks a lot for your investigations. You're right, I don't particularly need OTF, it was just the default format when downloading the fonts :)

Here are the `debug 2` logs, if this can help you :+1: 
[logs.txt](https://github.com/user-attachments/files/18285239/logs.txt)

**Comment by Finii** [comment]
The log looks excellent 🤔 Thanks for sharing.

I would propose to
* Ignore the `otf` problem for now (I fear that is a fontforge bug, and that will take a day or two to find)
* Check if the ligature problem can be solved by the new option from the PR
  * Let me check if the PR does not change anything unwanted
  * And then pull, a new docker image will be crated
  * You can check the real font

Is that a good plan?

Happy new year to you also 🎉 😃, thank you!

```console
$ grep -ve '\(^ *$\)\|^Updating\|^Adding' logs.txt
Running with options:
 --mono --debug 2 --careful --makegroups 2 --complete --has-no-italic --no-progressbars
Parallelism 0
Nerd Fonts Patcher v3.3.0 (4.16.2) (ff 20230101)
DEBUG: Naming mode 2
DEBUG: Monospace check: Panose says "monospaced"; glyph-width-mono True
INFO: Font vertical metrics slightly off (0.1%)
DEBUG: Font has negative right side bearing in extended glyphs
DEBUG: Final font cell dimensions 600 w x 1201 h (with icon cell 853 h)
DEBUG: 57/160 box drawing glyphs will be replaced
Done with Patch Sets, generating font...
DEBUG: Weight approximations: OS2/PS/Name: 400/400/400 (from 400/'Book'/'Regular')
DEBUG: =====> Family (ID 1)      ok       (19 <=31): TX02 Nerd Font Mono
DEBUG: =====> SubFamily (ID 2)   ok       ( 7 <=31): Regular
DEBUG: =====> Fullname (ID 4)    ok       (19 <=63): TX02 Nerd Font Mono
DEBUG: =====> PSN (ID 6)         ok       (11 <=63): TX02NFM-Reg
DEBUG: Tweaking 1/1
DEBUG: Changing flags from 0xB to 0x3
DEBUG: Changing lowestRecPPEM from 8 to 6
   TX02 Nerd Font Mono
   \===> '/out/TX02NerdFontMono-Regular.otf'
```

**Comment by Finii** [comment]
Please fetch a new docker font-patcher image.

The version tag should be `4.18.0` or later.

https://hub.docker.com/r/nerdfonts/patcher/tags

Then patch with these options:

```diff
 sudo docker run --rm \
   -v /home/clementpoiret/.local/share/fonts/TX-02:/in \
   -v /home/clementpoiret/.local/share/fonts/TX02-NerdFont:/out \
   nerdfonts/patcher \
   --progressbars \
-  --mono \
+  --single-width-glyphs \
   --adjust-line-height \
   -c
```

I.e. with the new `single-width-glyphs` that is a softer form of `mono`: It does not touch existing glyphs.
Maybe together with `--careful` which will prevent destroying preexisting glyphs by overpatching.

**Comment by thagabe** [comment]
Hello all! I bought into the Berkeley Fonts and noticed 2 things.

1. OTF - 2.3 MB per file which includes the ligatures (or at least the ligatures work) 
2. TTF - 24 kb per file and I don't think ligatures are included
3. Nerd Font Patcher works perfectly for TTF more or less (Exports BerkeleyMono Nerd Fonts) each file is about 4.7 MB
4. Nerd Font Patcher kinda sorts works with OTF but will only create a single file (multiple font types fail). It's pretty large and includes the ligatures but the weights are completely wrong. So far I have not been able to find any kind args that could fix this (I've tried --careful --adjusted-line-height --complete --single-width-glyph) the letters themselves seem to no longer have the weird offset but ligatures become tiny and/or shifted way to the left and most of it is hidden. I tried this on kitty, ghostty, macos terminal, and gnome terminal. 
5. Currently kitty loads glyphs from another source while using the OG Berkeley OTF font. But this does not work on different Terminals (ex. VS Code terminal). 

If possible could we re-open this issue?

**Comment by supermarin** [comment]
+1 got here noticing the same.

**Comment by Finii** [comment]
> If possible could we re-open this issue?

Sure!

For me the problem is that I can not directly examine the font files. I also considered buying them (I do buy fonts that I like), but that font is rather expensive and no font I would ever use. So I would spend that money just for this issue.

Then, I must admit I have forgotten what was the problem and what has been solved. So if you say 'me too' that can means a lot different things. At least attach images and kind-of fill the issue report (i.e. what OS, what has been done exactly, etc).

Also please give the in-parenthesis version of the patcher script, and possibly the version of `fontforge`. A screenshot or text capture of the patching is always good. Maybe add `--debug 2` to the patcher options.

> tried `--careful` `--adjusted-line-height` `--complete` `--single-width-glyph`

Explanations of the options are here: https://github.com/ryanoasis/nerd-fonts/wiki/ScriptOptions

Then I think, as this seems rather complex, we should focus on one font (one weight/style) first, e.g. `-Regular`. Only patch and install that one, test, then - if ok - remove the font, continue on with the next one. That minimizes inter-font-family problems and makes the issues more reproducible.

Well, how to continue? I guess I could write some 'diagnostics scripts' that one of you runs on the fonts and with the results maybe I get an idea what is going wrong. Not sure how effective that is. Or someone 'borrows' the problematic font file to me. I always respect font designers work and will neither use (nor forbid: share) the font, just analyze and try to patch. Afterwards I will give it back (i.e. erase on my machine). My email address is in every commit message.

But maybe we can gleam the reason from seeing the issues. Please provide images, with arrows probably pointing at the problem if it is subtle. Share patching output, run patching with `--debug 2`. Usually I run `fontforge font-patcher --debug 2 --complete SomeFont.otf 2>/dev/null` to ignore the fontforge output. Ah, do you run the docker, or direct, etc? Best is to not-use the docker version because there are some problems with forwarding the error messages (or did I fix it?).

Ah, - no image 😬 - but if the ligatures are too small and the rest is ok: Is that the Nerd Font Mono (`--mono`) variant? Ligatures can be formulated in so many different ways, and the usual 'ideoms' are supported by the patcher. For example _Monaspace_ also had a Ligature problem with patching, but that was in effect a problem in the original font, and I believe they changed the way the ligatures are 'spelled out' in the font files. "Normal" fonts, and what font designers are usually working with, have a straight forward way for ligatures. But that breaks on terminal use, and a way that feels strange (if you never did terminal fonts) must be used (one recent post that turned up on Hacker News: [1]).

Well. Too long already.

Cheers,  Fini


[1] https://joshleeb.com/posts/monospace-ligatures.html

**Comment by Finii** [comment]
>  Usually I run `fontforge font-patcher --debug 2 --complete SomeFont.otf 2>/dev/null` to ignore the fontforge output. 

Ah, obviously I added a new option just for Berkeley, and you need to give `--single-width-glyphs` instead of `--mono`; if you target a _Nerd Font Mono_ font variant.

That resonates with _but ligatures become tiny and/or shifted way to the left_.
Hmm. Well, I mostly poke in the dark here, you need to share concrete information with images ;)

<img width="933" height="409" alt="Image" src="https://github.com/user-attachments/assets/5b98d89a-b4ba-4105-823d-eb155039b7b2" />

**Comment by doubleyewdee** [comment]
> 4. Nerd Font Patcher kinda sorts works with OTF but will only create a single file (multiple font types fail). It's pretty large and includes the ligatures but the weights are completely wrong. So far I have not been able to find any kind args that could fix this (I've tried --careful --adjusted-line-height --complete --single-width-glyph) the letters themselves seem to no longer have the weird offset but ligatures become tiny and/or shifted way to the left and most of it is hidden. I tried this on kitty, ghostty, macos terminal, and gnome terminal.

This is the behavior I noticed as well that led me here. I am not font-savvy at all, but inspected the "SuperType Compiler"-emitted OTF files in FontForge. One thing I noticed is that, in the TrueType metadata (`TTF Names`), the `Styles (SubFamily)` field for all OTF files is the same, `Regular`. My uninformed guess is that FontForge is keying off that value somehow? Metadata (clipped license data at bottom):

<img width="2306" height="1158" alt="Image" src="https://github.com/user-attachments/assets/944ef161-6dba-4db2-89d3-a0ddee1e7512" />

However, I also noticed that, in FontForge, the variant I am inspecting ("Bold Oblique") has some *very* misaligned characters. So it didn't seem worthwhile to edit and re-export to test my theory. Sample (note the C, T, Y characters, in addition to others):

<img width="1165" height="518" alt="Image" src="https://github.com/user-attachments/assets/cf41f8a9-7e6d-4939-951f-cf2e12b8d75a" />

I'm not sure if getting TTF vs OTF variants would be helpful to me in terms of the converter, though I will try it after responding here and see how it goes. I like the ligatures quite a lot, but it sounds like I might either lose them or suffer degraded quality by going from OTF->TTF?

Lastly, @Finii, I am happy to cover the cost of the font license for you to assist in debugging this on your end, if that is something you'd be interested in. I'm way out of my depth here, but I adore this font, and would like to get the extended Nerd Fonts glyphs working with it. :)

**Comment by Finii** [comment]
Are the fontforge images of the font before or after patching? The name does not have "Nerd Font" in it 🤔 

> This is the behavior I noticed as well that led me here. I am not font-savvy at all, but inspected the "SuperType Compiler"-emitted OTF files in FontForge. One thing I noticed is that, in the TrueType metadata (`TTF Names`), the `Styles (SubFamily)` field for all OTF files is the same, `Regular`. My uninformed guess is that FontForge is keying off that value somehow? 

Fontforge is working fine in that regard. If that is post-Nerd-patching ... The patcher script needs to understand the naming of the font to come up with correctly grouped new names that include 'Nerd Font' or 'NF'. Maybe it can not determine the original weight and thus assumes 'Regular'. It would be helpful if you can post the `font-patcher` output if run with `--debug 2` - there will be information on the weight and possible problems with the different places a font has for the weight.

\<image\>

If that is after patching this is really wrong and should not have happened. But there is something amiss with the naming anyhow, as there is no "Nerd Font" or "NF" in the name? If the renaming is turned off it will/should not touch the names (weights etc) at all. 🤔 

> However, I also noticed that, in FontForge, the variant I am inspecting ("Bold Oblique") has some _very_ misaligned characters.

I'm not sure how well `font-patcher` handles Oblique (and there are bugs handling Slanted). Best would be to look at other fonts first.

> I'm not sure if getting TTF vs OTF variants would be helpful to me in terms of the converter, though I will try it after responding here and see how it goes. I like the ligatures quite a lot, but it sounds like I might either lose them or suffer degraded quality by going from OTF->TTF?

Quoting from above:
* The ttf is also an OpenType font, nowadays ;) and it has technically more capabilities. If I have a choice here we usually patch (only) the ttf variants.
* While the otf has the blues defined the ttf seems to be not even autohinted
* OTF - 2.3 MB per file which includes the ligatures (or at least the ligatures work) TTF - 24 kb per file and I don't think ligatures are included
* The shifted letters like r seem to be a fontforge bug. Built HEAD from git, and it is still there.

So it seems we need to patch the otf fonts. The fontforge bug needs to be fixed first, of course. Did fix already some bugs upstream (i.e. at fontforge), but that fix can be difficult or easy, you never know. Hmm, very disappointing.

> Lastly, [@Finii](https://github.com/Finii), I am happy to cover the cost of the font license for you to assist in debugging this on your end, if that is something you'd be interested in. I'm way out of my depth here, but I adore this font, and would like to get the extended Nerd Fonts glyphs working with it. :)

It should be possible to fix it if I get my hands on the real fonts. 

<img width="814" height="385" alt="Image" src="https://github.com/user-attachments/assets/3bcb11e1-b7e7-422d-8903-2ff9d26df9f4" />

Ah, quoting myself from this comment above

> Are the fontforge images of the font before or after patching?

This is the output of their "Compiler", i.e. 'original' font.

* Do the Compiler fonts differ from the stand-alone fonts? I have seen problems with 'font customizer' stuff in the past; there are several (!) bugs in [opentype JS](https://opentype.js.org/), and altough I PR'ed fixes the library is still broken (fixes not merged) 😒 
* If I read that correctly it is 75+20 for the font with ligatures?

So, _I am happy to cover the cost of the font license for you to assist in debugging this on your end, if that is something you'd be interested in._ There are several problems with this font I guess. The displaced characters are also in the example font, so the fontforge bug could be tackled with the example font files. The naming ... there could be a problem in the (my) `FontnameParser`, esp Oblique and Slanted is not very well supported. To find problems with the names of the original or patched fonts we could use the `bin/scripts/name_parser/query_*` scripts on your machine.
If the ligature problem goes away with `--single-width-glyphs` there seems to be no real need for the font files, except that it's of course much easier than 'remote debugging'.

Also interesting if you want a `Nerd Font` font or a `Nerd Font Mono` font; noone seems to have that answered above. There is not a good documentation about the variants, but maybe this helps: https://github.com/ryanoasis/nerd-fonts/discussions/1103

**Comment by supermarin** [comment]
Update from my end: spoke to Neil from US Graphics and he advised to go with .ttf. 
Since switching, everything looks good.
@Finii you might want to reach out to Neil and see if you could get a developer license for the font.

**Comment by Finii** [comment]
Finding from #1942:

Use `--name postscript` for the `otf` fonts.
Ligs may be broken there.

**Comment by GreenOolong** [comment]
./font-patcher --careful --complete --progressbars --single-width-glyphs --adjust-line-height --outputdir . --name postscript file_path not working

**Comment by Finii** [comment]
@ChihGodlee What does _not working_ mean? Is there a message or show an image of what goes wrong?

**Comment by GreenOolong** [comment]
> [@ChihGodlee](https://github.com/ChihGodlee) What does _not working_ mean? Is there a message or show an image of what goes wrong?

I read the comments above that everything worked fine after using these parameters, but I'm still getting weird misalignments when I use these parameters.

**Comment by Finii** [comment]
> I read the comments above that everything worked fine after using these parameters, but I'm still getting weird misalignments when I use these parameters.

An image would always be best. And here, are you sure that is not a previous patchversion cached by the OS (in case of Windows or MacOs), did you do the propper font 'hygiene'? All information asked for a new issue could help. The `--debug 2` output and the real command line you used would help. I would need to reproduce the problem which is difficult with the minimal information provided by you.

**Comment by GreenOolong** [comment]
> > I read the comments above that everything worked fine after using these parameters, but I'm still getting weird misalignments when I use these parameters.
> 
> An image would always be best. And here, are you sure that is not a previous patchversion cached by the OS (in case of Windows or MacOs), did you do the propper font 'hygiene'? All information asked for a new issue could help. The `--debug 2` output and the real command line you used would help. I would need to reproduce the problem which is difficult with the minimal information provided by you.

unpatched:
<img width="114" height="66" alt="Image" src="https://github.com/user-attachments/assets/fac3cba5-01a7-4ba6-8b6a-3edb2ddb1730" />

patched:
<img width="116" height="56" alt="Image" src="https://github.com/user-attachments/assets/0926adbc-d25a-41da-9635-eb526dc598da" />

Maybe the problem has something to do with ligature? I've tested it and it seems to work fine without the ligature.

[debug.txt](https://github.com/user-attachments/files/24285458/debug.txt)

**Comment by Finii** [comment]
@ChihGodlee can reproduce

**Comment by Finii** [comment]
Not sure what happens here, but it seems a `fontforge` bug. I opened and exported unchanged (no font-patcher script involved), and the left side bearing doubles for each and every glyph?!

<img width="1920" height="658" alt="Image" src="https://github.com/user-attachments/assets/f93339fb-ed3f-44e9-8222-42a44485641c" />

This is ff 20230101, lets try the latest release...

**Comment by Finii** [comment]
That has been fixed in `fontforge` ~~20251009~~ `HEAD`, lets see if I can bisect it.

It has been fixed by fontforge commit [9d119ca0bf415e3d69086f4931fcbb39d3d85c51](https://github.com/fontforge/fontforge/commit/9d119ca0bf415e3d69086f4931fcbb39d3d85c51)

This PR fixed the problem:

* https://github.com/fontforge/fontforge/pull/5698

So... Sorry, we can do nothing here to address the problem because it is a `fontforge` bug which has been fixed but it is not yet released. So you need your own build of `fontforge`, or for example Arch's package.

Note that you also need [aedb8f2e95ddba99e8cb6446f082a356e2c2625b](https://github.com/fontforge/fontforge/commit/aedb8f2e95ddba99e8cb6446f082a356e2c2625b) if you want to generate a `Nerd Font Mono` font (which fixes a bug introduced with `fontforge` 20251009.

**Comment by Finii** [comment]
Hmm, even Arch's is not enough, they just take the release 20251009 and add my fix, not the bit-1 fix.

Out docker uses Alpine latest-stable which is 20230101 :-(

Maybe we have to build on our own, for the docker and for the release process as well ...

<img width="1185" height="809" alt="Image" src="https://github.com/user-attachments/assets/3c57eea6-c551-4582-9067-951b780dd4d5" />

_PKGBUILD with release and one backport PR_

**Comment by GreenOolong** [comment]
Well, I don't quite understand what I should do.

**Comment by Finii** [comment]
> Well, I don't quite understand what I should do.

You need a bleeding edge `fontforge`, newer than the newest release.

I do not know where you can get one pre-built, but building it on your own is not too hard - on Linux.

You can follow the instructions given here:
https://github.com/fontforge/fontforge/blob/master/INSTALL.md

You do not need to do the `sudo ninja install` step, the built executabkle also runs uninstalled, like (example)

```console
$ git/fontforge/build/bin/fontforge git/nerd-fonts/font-patcher --complete --name postscript .....
```

This is only true/needed for the `.otf` fonts. If you can live without ligatures just patch and use the `.ttf` version of Berkeley Mono which does not have all the problems and patches fine with any recent `fontforge` release version.

_Edit:_

If you can use docker - I think about putting the HEAD `fontforge` into the "nerdfonts/patcher" docker image. But that will take some time and will most likely come only later this week (if at all).

https://hub.docker.com/r/nerdfonts/patcher

**Comment by Eplankton** [comment]
> > Well, I don't quite understand what I should do.
> 
> You need a bleeding edge `fontforge`, newer than the newest release.
> 
> I do not know where you can get one pre-built, but building it on your own is not too hard - on Linux.
> 
> You can follow the instructions given here: https://github.com/fontforge/fontforge/blob/master/INSTALL.md
> 
> You do not need to do the `sudo ninja install` step, the built executabkle also runs uninstalled, like (example)
> 
> $ git/fontforge/build/bin/fontforge git/nerd-fonts/font-patcher --complete --name postscript .....
> 
> This is only true/needed for the `.otf` fonts. If you can live without ligatures just patch and use the `.ttf` version of Berkeley Mono which does not have all the problems and patches fine with any recent `fontforge` release version.
> 
> _Edit:_
> 
> If you can use docker - I think about putting the HEAD `fontforge` into the "nerdfonts/patcher" docker image. But that will take some time and will most likely come only later this week (if at all).
> 
> https://hub.docker.com/r/nerdfonts/patcher

Hello there, will we finally get a completely fixed tools for `.otf` format?

**Comment by Finii** [comment]
As that is a `fontforge` bug we can do nothing about it. Unfortunately a fully fixed `fontforge` has not been released yet.

To help out we build `fontforge` from scratch for you in our docker image.

Please use that, see

* https://hub.docker.com/r/nerdfonts/patcher
* https://github.com/ryanoasis/nerd-fonts#font-patcher (scroll a bit down to "use docker")

Remember to update your font-patcher image. `docker images` shows you your images. The newest at the moment are

<img width="719" height="205" alt="Image" src="https://github.com/user-attachments/assets/172e2fa9-c78d-41e5-80a7-b40854bf8cfb" />
